src/libfsm/determinise.c

```diff
@@ -284,109 +284,171 @@ nextnotdone(struct mapping_set *mappings, struct mapping_iter_save *sv)
 	return NULL;
 }
 
-/*
- * List all states within a set which are reachable via non-epsilon
- * transitions (that is, have a label associated with them).
+static void
+carryend(struct state_set *set, struct fsm *fsm, struct fsm_state *state)
+{
+	struct state_iter it;
+	struct fsm_state *s;
+
+	assert(set != NULL); /* TODO: right? */
+	assert(fsm != NULL);
+	assert(state != NULL);
+
+	for (s = state_set_first(set, &it); s != NULL; s = state_set_next(&it)) {
+		if (fsm_isend(fsm, s)) {
+			fsm_setend(fsm, state, 1);
+		}
+	}
+}
+
+static int
+cmp_single_state(const void *a, const void *b)
+{
+	const struct fsm_state *ea = a, *eb = b;
+	return (ea > eb) - (ea < eb);
+}
+
+static unsigned long
+hash_single_state(const void *a)
+{
+	const struct fsm_state *st = a;
+	return hashrec(st, sizeof *st);
+}
+
+/* builds transitions from current dfa state to new dfa states
+ *
+ * Makes a single pass through each of the nfa states that this dfa state
+ * corresponds to and constructs an edge set for each symbol out of the nfa
+ * states.
  *
- * TODO: maybe simpler to just return the set, rather than take a double pointer
+ * It does this in one pass by maintaining an array of destination states where
+ * the index of the array is the transition symbol.
  */
 static int
-listnonepsilonstates(const struct fsm *fsm, struct trans_set *trans, struct state_set *set)
+buildtransitions(const struct fsm *fsm, struct fsm *dfa, struct mapping_set *mappings, struct mapping *curr)
 {
 	struct fsm_state *s;
 	struct state_iter it;
+	int sym, sym_min, sym_max;
+	int ret = 0;
 
-	assert(set != NULL);
-	assert(trans != NULL);
+	struct state_array outedges[FSM_SIGMA_COUNT];
+	struct hashset *outsets[FSM_SIGMA_COUNT];
```

Comment:
If we included the symbol in the hash function, could we use the same hashset for all edges here? Would that be worse?

Reply:
I'm not sure.  That's worth benchmarking.